src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs

```diff
@@ -169,7 +169,17 @@ public CachingContext(JsonSerializerOptions options, int hashCode)
             // If changing please ensure that src/ILLink.Descriptors.LibraryBuild.xml is up-to-date.
             public int Count => _jsonTypeInfoCache.Count;
 
-            public JsonTypeInfo? GetOrAddJsonTypeInfo(Type type) => _jsonTypeInfoCache.GetOrAdd(type, Options.GetTypeInfoNoCaching);
+            public JsonTypeInfo? GetOrAddJsonTypeInfo(Type type)
+            {
+#if NETCOREAPP2_0_OR_GREATER
+                return _jsonTypeInfoCache.GetOrAdd(type, static (type, options) => options.GetTypeInfoNoCaching(type), Options);
+#else
+                return _jsonTypeInfoCache.TryGetValue(type, out JsonTypeInfo? value) ?
+                    value :
+                    _jsonTypeInfoCache.GetOrAdd(type, Options.GetTypeInfoNoCaching(type));
+#endif
+            }
+
```

Comment:
Couldn't we just avoid duplicate lookups in ns2.0 by caching the delegate?
```suggestion
            private readonly Func<Type, JsonTypeInfo> _addDelegate = type => Options.GetTypeInfoNoCaching(type);
            
            public JsonTypeInfo? GetOrAddJsonTypeInfo(Type type) => _jsonTypeInfoCache.GetOrAdd(type, _addDelegate);

```

Reply:
You'd end up with a delegate allocation per CachingContext.  If that doesn't matter, sure, we could do that instead (though not exactly as in the proposed code, since you can't access an instance member in the initialization for a field).

Reply:
If we cared much we could still use the stateful overload in netcoreapp targets, but I suspect it wouldn't do enough to justify the added complexity.